[WIP]: Project node selector feature by using taints and tolerations

aveshagarwal opened the pull request:
I have to really polish/refactor/test properly. But it compiles and unit tests pass.

aveshagarwal commented:
@ncdc

derekwaynecarr commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	nsName := a.GetNamespace()
+	var namespace *api.Namespace
+
+	namespaceObj, exists, err := p.store.Get(&api.Namespace{
```
Just a heads up but will want to move this to use shared informer as is done here:

https://github.com/kubernetes/kubernetes/pull/29926

derekwaynecarr replied:
Also add the wait for ready checks

derekwaynecarr commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+
+func NewPodNodeEnvironment(client clientset.Interface, clusterDefaultTolerations []api.Toleration) *podNodeEnvironment {
+	// TODO: make it a shared cache to use between admission plugins
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
```
Ok, before we open the main pull make the switch 😜

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
```
Wasn't "All rights reserved" deprecated? I was of the impression that this even fails the tests.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	client    clientset.Interface
+	store     cache.Store
+	reflector *cache.Reflector
+	// global default tolerations in a cluster, If a namespace is
```
nit: , => .

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	})
+}
+
+// podNodeEnvironment is an implementation of admission.Interface.
```
better than any comment: `var _ admission.Interface = podNodeEnvironment{}`

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+type pluginConfig struct {
+	PodNodeEnvironmentPluginConfig []api.Toleration
```
strange identifier name. Either s/PodNodeEnvironmentPluginConfig/tolerations or

``` golang
PodNodeEnvironmentPluginConfig struct {
  tolerations []api.Toleration
}
```

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+
+// readConfig reads default value of clusterDefaultTolerations
+// from the file provided with --admission-control-config-file
+// If the file is not supplied, it defaults to ""
```
nit: full-stop

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+// If the file is not supplied, it defaults to ""
+// The format in a file:
+// podNodeEnvironmentPluginConfig:
+//  clusterDefaultTolerations: <tolerations>
```
see comment above about the config. Seems like the comment matches what I suggested.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		glog.Errorf("expected pod but got %s", a.GetKind().Kind)
+		return nil
```
isn't this an error and not nil?

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+type podNodeEnvironment struct {
+	*admission.Handler
+	client    clientset.Interface
+	store     cache.Store
```
better name: `nsStore` or just `namespaces`

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	return defaultConfig
+}
+
+// Admit merges pod's and its namespace tolerations.
```
better: Admit merges tolerations of the pod with those of the podNodeEnvironment

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+func (p *podNodeEnvironment) defaultGetNamespace(name string) (*api.Namespace, error) {
+	namespace, err := p.client.Core().Namespaces().Get(name)
+	if err != nil {
+		return nil, fmt.Errorf("namespace %s does not exist", name)
```
it can be any other error as well. I wouldn't decorate it, just return err.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	}
+}
+
+func (p *podNodeEnvironment) defaultGetNamespace(name string) (*api.Namespace, error) {
```
feels odd to call this `defaultGetNamespace`, but then pass a name. Why not call `p.client.Core().Namespaces().Get(name)` directly? Doesn't seem to have added value.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	return namespace, nil
+}
+
+func (p *podNodeEnvironment) getNameSpaceTolerations(namespace *api.Namespace) ([]api.Toleration, error) {
```
s/NameSpace/Namespace/

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+// conflictingTolerations returns true if the key of two tolerations match
+// but one or more other fields differ, otherwise returns false
+func conflictingTolerations(first []api.Toleration, second []api.Toleration) bool {
+	for i := range first {
```
quadratic complexity, but probably it's still better than a n*log(n) algorithm with few tolerations.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+	for i := range first {
+		for j := range second {
+			if first[i].Key == second[j].Key {
+				if first[i].Operator != second[j].Operator ||
```
Isn't there some (partial) order of tolerations and merging should take the maximal toleration? E.g.:
- namespace: `<"env", "prod", "">`, equal
- pod: `<"env", "", "">`, exists

This means that all pods in the namespace tolerate "env=prod" taints. But the given pod even tolerates any "env" taints, e.g. also "env=test".

aveshagarwal replied:
I am still thinking about it, so mostly in next update.

sttts commented on plugin/pkg/admission/nodeenv/admission.go:
```diff
@@ -0,0 +1,234 @@
+// it does not check for conflicts
+// it assumes no duplicates in individual tolerations
+func mergeTolerations(first []api.Toleration, second []api.Toleration) []api.Toleration {
+	var mergedTolerations []api.Toleration
```
make([]api.Toleration, 0, len(second))

aveshagarwal replied:
is there any advantage of doing it explicitly, may be I am not aware?

sttts replied:
You avoid some memory allocator work if you know the size in advance.

Commit:
Project node selector feature by using taints and tolerations.

aveshagarwal commented:
@derekwaynecarr @sttts thanks for your feedback. Although I have updated this PR, but it does not yet addresses all your comments. Will address the remaining comments today.

ingvagabund commented on pkg/api/helpers.go:
```diff
@@ -449,9 +449,9 @@ func GetAffinityFromPodAnnotations(annotations map[string]string) (*Affinity, er
+// GetTolerationsFromAnnotations gets the json serialized tolerations data from Annotations
 // and converts it to the []Toleration type in api.
-func GetTolerationsFromPodAnnotations(annotations map[string]string) ([]Toleration, error) {
+func GetTolerationsFromAnnotations(annotations map[string]string) ([]Toleration, error) {
```
You are renaming exported symbol. Are you sure the symbol is not used anywhere else? Even in projects importing `k8s.io/kubernetes/pkg/api` package?

aveshagarwal replied:
I changed it already wherever it was used in other places. Is there something I am missing?

ncdc replied:
It's possible that downstream code (OpenShift) could use it. I'm comfortable with the rename, however, and if OpenShift does use it, we'll just fix it in a rebase.

aveshagarwal commented:
@derekwaynecarr @sttts implemented shared informer cache and incorporated feedback. PTAL.

sttts commented on plugin/pkg/admission/podplacementrestriction/admission.go:
```diff
@@ -0,0 +1,248 @@
+		return err
+	}
+
+	if conflictingTolerations(nsTolerations, podTolerations) {
```
still thinking this should be more clever in allowing the maximum toleration in the merge. If that does not exist, then it's a conflict.

aveshagarwal replied:
@sttts yeah I haven't thought much about it, basically replicating what existed with levels. I think will give it a more thought now.

aveshagarwal commented:
I have sent a PR to kube upstream so closing this. Lets discuss there. I have not made conflict and merging methods more intelligent yet.
